### PR TITLE
feat: Convergence Early Stopping Add adaptive convergence criteria and utilities

### PR DESCRIPTION
**Pull request by dferguson992**
## Summary
This PR introduces adaptive convergence and detailed aggregation to AIPerf's multi-run orchestration. Instead of always running a fixed number of trials, users can now specify a convergence criterion that stops benchmarking early once metrics stabilize. A new detailed aggregation strategy reads per-request JSONL data to compute true combined percentiles across all runs, complementing the existing confidence-interval aggregation. This builds on top of the multi-run orchestration framework introduced in prior PRs.

## Core Features

- Convergence Criterion Framework: Pluggable ConvergenceCriterion base class with shared JSONL loader that filters warmup records, handles malformed lines, and skips missing files gracefully
- CI Width Convergence (ci_width): Stops when the Student's t confidence interval width relative to the mean falls below a threshold. Operates on run-level summary stats for any metric/statistic combination
- CV Convergence (cv): Stops when the coefficient of variation (std/mean) across run-level averages drops below a threshold
- Distribution Convergence (distribution): Uses a two-sample Kolmogorov-Smirnov test on per-request JSONL data to detect when the latest run's distribution matches prior runs, catching bimodal behavior and tail shifts that summary statistics miss
- Adaptive Execution Strategy: Composes with any convergence criterion, respects configurable min/max run bounds, and mirrors FixedTrialsStrategy's run labeling, path structure, seed handling, and warmup disabling for full artifact compatibility
- Detailed Aggregation: Reads per-request JSONL from all runs, combines into a single population per metric, and computes true combined percentiles (p50/p90/p95/p99), mean, std, and per-run breakdowns
- Configuration & CLI
- New flags: --convergence-metric, --convergence-mode, --convergence-threshold, --convergence-stat
- Validation: convergence requires --num-profile-runs > 1; distribution mode requires JSONL export level
- Backward compatible: no convergence flags = FixedTrialsStrategy, identical to previous behavior

## Exports
New profile_export_aiperf_detailed.json in the aggregate directory when adaptive mode is used
Existing profile_export_aiperf_aggregate.json and CSV continue to be produced

## Documentation
CLI options reference updated with all four convergence flags

## Test Coverage
17 new test files, ~2900 lines added
Unit tests for all three convergence criteria covering: below min_runs, converged/not-converged states, all-failed-runs, missing metrics, boundary conditions, warmup exclusion
Unit tests for AdaptiveStrategy: early stopping, max_runs cap, min_runs floor, seed handling, warmup disabling
Unit tests for DetailedAggregation: combined percentiles, per-run breakdowns, mixed success/failure, malformed JSONL, multiple metrics, IO errors
CLI wiring tests verifying correct strategy/criterion creation from flags
Integration test using AIPerfCLI + AIPerfMockServer verifying adaptive early stopping, backward compat, CV mode, and request-rate cross-mode compatibility

## Example
Command to trigger Adaptive Strategy's early termination:
```
aiperf profile \
  --model openai/gpt-oss-20b \
  --url http://localhost:8080 \
  --endpoint-type chat \
  --num-requests 20 \
  --concurrency 1 \
  --num-profile-runs 10 \
  --convergence-mode ci_width \
  --convergence-threshold 0.5 \
  --convergence-metric output_token_throughput \
  --convergence-stat avg \
  --use-server-token-count
```

Output from early Adaptive Strategy, converged on round 3 of 20
```
16:47:26.412 INFO     ================================================================================ (cli_runner.py:383)
16:47:26.412 INFO     AGGREGATE STATISTICS SUMMARY (cli_runner.py:384)
16:47:26.413 INFO     ================================================================================ (cli_runner.py:385)
16:47:26.413 INFO     Aggregation Type: confidence (cli_runner.py:386)
16:47:26.413 INFO     Total Runs: 3 (cli_runner.py:387)
16:47:26.413 INFO     Successful Runs: 3 (cli_runner.py:388)
16:47:26.413 INFO     Confidence Level: 95% (cli_runner.py:397)
16:47:26.413 INFO      (cli_runner.py:399)
16:47:26.413 INFO     Key Metrics: (cli_runner.py:400)
16:47:26.413 INFO     -------------------------------------------------------------------------------- (cli_runner.py:401)
16:47:26.414 INFO     
Request Throughput (Avg): (cli_runner.py:435)
16:47:26.414 INFO       Mean:          0.2069 requests/sec (cli_runner.py:436)
16:47:26.414 INFO       Std Dev:       0.0053 requests/sec (cli_runner.py:437)
16:47:26.414 INFO       Min:           0.2021 requests/sec (cli_runner.py:438)
16:47:26.414 INFO       Max:           0.2125 requests/sec (cli_runner.py:439)
16:47:26.414 INFO       CV:             2.54% (cli_runner.py:440)
16:47:26.414 INFO       95% CI: [0.1938, 0.2199] requests/sec (cli_runner.py:441)
16:47:26.415 INFO     
Request Latency (Avg): (cli_runner.py:435)
16:47:26.415 INFO       Mean:       4834.8590 ms (cli_runner.py:436)
16:47:26.415 INFO       Std Dev:     122.2312 ms (cli_runner.py:437)
16:47:26.415 INFO       Min:        4704.4016 ms (cli_runner.py:438)
16:47:26.415 INFO       Max:        4946.7367 ms (cli_runner.py:439)
16:47:26.415 INFO       CV:             2.53% (cli_runner.py:440)
16:47:26.415 INFO       95% CI: [4531.2198, 5138.4982] ms (cli_runner.py:441)
16:47:26.416 INFO      (cli_runner.py:449)
16:47:26.416 INFO     -------------------------------------------------------------------------------- (cli_runner.py:450)
16:47:26.416 INFO     Coefficient of Variation (CV) Interpretation Guide: (cli_runner.py:451)
16:47:26.416 INFO       CV < 5%:   Excellent repeatability (low variance) (cli_runner.py:452)
16:47:26.416 INFO       CV 5-10%:  Good repeatability (moderate variance) (cli_runner.py:453)
16:47:26.416 INFO       CV 10-20%: Fair repeatability (consider more runs) (cli_runner.py:454)
16:47:26.416 INFO       CV > 20%:  High variance (investigate or increase runs) (cli_runner.py:455)
16:47:26.417 INFO      (cli_runner.py:456)
16:47:26.417 INFO     Confidence Interval (CI) Interpretation: (cli_runner.py:457)
16:47:26.417 INFO       The 95% CI indicates the range where the true mean (cli_runner.py:458)
16:47:26.417 INFO       is likely to fall with 95% confidence. (cli_runner.py:461)
16:47:26.417 INFO       Narrower intervals indicate more precise estimates. (cli_runner.py:462)
16:47:26.417 INFO     ================================================================================ (cli_runner.py:463)
```

## Future Work
Dashboard: TUI visualization of convergence progress across runs
Additional convergence criteria (e.g., relative change, custom plugin criteria)

## Summary by CodeRabbit

* **New Features**
  * Adaptive convergence for multi-run benchmarks (CI-width, CV, distribution) with early stopping, new adaptive strategy, and an optional detailed aggregate JSON export.
  * New detailed aggregation path exposing per-run distributions and breakdowns.

* **Documentation**
  * New CLI options: --convergence-metric, --convergence-stat, --convergence-threshold, --convergence-mode.

* **Tests**
  * Extensive unit and integration tests covering convergence criteria, adaptive strategy, detailed aggregation, and exporter behavior.

* **Chores**
  * Build scripts: PATH assignments now quoted for robust handling of spaces.

### CODE REVIEW

File: src/aiperf/exporters/aggregate/aggregate_detailed_json_exporter.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    def get_file_name(self) -> str:
+        return "profile_export_aiperf_detailed.json"
+
+    def _generate_content(self) -> str:
+        from importlib.metadata import version as get_version
+
+        try:
+            aiperf_version = get_version("aiperf")
+        except Exception:
+            aiperf_version = "unknown"
```
Comment (ajcasagrande) [inline src/aiperf/exporters/aggregate/aggregate_detailed_json_exporter.py]: I think I standardized on doing this once in the `__init__.py`, so you should be able to do `from aiperf import __version__ as aiperf_version`. You can look at some other files

**Comment by ajcasagrande** [comment]
@coderabbitai can you confirm the following: https://github.com/ai-dynamo/aiperf/pull/761#issuecomment-4072562455

@dferguson992 can you please take a look and confirm if any of these are real issues or not? thanks.

I appreciate the PR!

File: src/aiperf/exporters/aggregate/aggregate_detailed_json_exporter.py (hunk 1 above)
Comment (dferguson992) [reply src/aiperf/exporters/aggregate/aggregate_detailed_json_exporter.py]: Changed to `from aiperf import __version__ as aiperf_version` in the next commit.

File: src/aiperf/orchestrator/aggregation/detailed.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+"""Detailed aggregation strategy using per-request JSONL data."""
+
+import numpy as np
+
+from aiperf.orchestrator.aggregation.base import AggregateResult, AggregationStrategy
+from aiperf.orchestrator.jsonl_loader import DEFAULT_JSONL_FILENAME, load_all_metrics
+from aiperf.orchestrator.models import RunResult
+
+
+class DetailedAggregation(AggregationStrategy):
```
Comment (Lokiiiiii) [inline src/aiperf/orchestrator/aggregation/detailed.py:12]: Is this required for any of the convergence strategies ? Usually we want to keep the confidence trials IIDs. Aggregating across IID runs is unconventional AFAIK

Comment (dferguson992) [reply src/aiperf/orchestrator/aggregation/detailed.py:12]: No `DetailedAggregation` is not used by any convergence strategy. Convergence decisions use only `RunResult.summary_metrics` (ci_width/cv) or per-run JSONL independently.

`DetailedAggregation` is purely a post-hoc reporting step: after all runs complete, it pools per-request records to compute combined percentiles over the full population. It doesn't feed back into the convergence loop. The IID property of the runs is preserved and convergence still operates on independent run-level statistics.

File: src/aiperf/common/config/loadgen_config.py [hunk 3]
```diff
@@ -510,6 +572,58 @@ def validate_multi_run_params(self) -> "LoadGeneratorConfig":
...
+            "convergence_threshold": "--convergence-threshold",
+            "convergence_stat": "--convergence-stat",
+        }
+        if self.convergence_metric is None:
+            for field_name, flag_name in convergence_dependent_flags.items():
+                if field_name in self.model_fields_set:
+                    raise ValueError(
+                        f"{flag_name} requires --convergence-metric to be set. "
+                        f"Either add --convergence-metric or remove {flag_name}."
+                    )
```
Comment (ajcasagrande) [inline src/aiperf/common/config/loadgen_config.py]: Bug: duplicate block. Lines 601-625 are an exact copy of lines 588-599 + 577-585 above.

The `convergence_dependent_flags` dict and loop appear twice, and the `convergence_stat` + distribution check appears twice. The second copy of the distribution check (line 619) also compares against the string `"distribution"` instead of `ConvergenceMode.DISTRIBUTION`.

Delete this entire duplicate block (lines 601-625).

**Review by ajcasagrande — COMMENTED** [review]
Two issues spotted. Overall looking good, though.
